pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
+		client:  iamClient,
+		filters: o.Filters,
+		logger:  o.Logger,
+	}
+
+	var loopError error
+	for len(tagClients) > 0 || loopError != nil {
+		loopError = nil
+		nextTagClients := tagClients[:0]
+		for _, tagClient := range tagClients {
+			matched := false
+			for _, filter := range o.Filters {
+				o.Logger.Debugf("search for and delete matching resources by tag in %s matching %#+v", tagClientNames[tagClient], filter)
+				tagFilters := make([]*resourcegroupstaggingapi.TagFilter, 0, len(filter))
+				for key, value := range filter {
+					tagFilters = append(tagFilters, &resourcegroupstaggingapi.TagFilter{
+						Key:    aws.String(key),
+						Values: []*string{aws.String(value)},
+					})
+				}
+				err = tagClient.GetResourcesPages(
+					&resourcegroupstaggingapi.GetResourcesInput{TagFilters: tagFilters},
+					func(results *resourcegroupstaggingapi.GetResourcesOutput, lastPage bool) bool {
+						for _, resource := range results.ResourceTagMappingList {
+							arn := *resource.ResourceARN
+							if _, ok := deleted[arn]; !ok {
+								matched = true
+								err := deleteARN(awsSession, arn, o.Logger)
+								if err == nil {
+									deleted[arn] = exists
+								} else {
+									err = errors.Wrapf(err, "deleting %s", arn)
+									o.Logger.Debug(err)
+								}
+							}
+						}
+
+						return !lastPage
+					},
+				)
+				if err != nil {
+					err = errors.Wrapf(err, "get tagged resources")
+					o.Logger.Info(err)
+					loopError = err
+				}
+			}
+
+			if matched {
+				nextTagClients = append(nextTagClients, tagClient)
+			} else {
+				o.Logger.Debugf("no deletions from %s, removing client", tagClientNames[tagClient])
+			}
 		}
-	}
+		tagClients = nextTagClients
+
+		o.Logger.Debug("search for IAM roles")
+		arns, err := iamRoleSearch.arns()
+		if err != nil {
+			o.Logger.Info(err)
+			loopError = err
+		}
+
+		o.Logger.Debug("search for IAM users")
+		userARNs, err := iamUserSearch.arns()
+		if err != nil {
+			o.Logger.Info(err)
+			loopError = err
+		}
+		arns = append(arns, userARNs...)
 
-	// wait for them to finish
-	for goroutines > 0 {
-		select {
-		case res := <-returnChannel:
-			goroutines--
-			o.Logger.Debugf("goroutine %v complete (%d left)", res, goroutines)
+		if len(arns) > 0 {
+			o.Logger.Debug("delete IAM roles and users")
+		}
+		for _, arn := range arns {
+			if _, ok := deleted[arn]; !ok {
+				err = deleteARN(awsSession, arn, o.Logger)
+				if err == nil {
+					deleted[arn] = exists
+				} else {
+					err = errors.Wrapf(err, "deleting %s", arn)
+					o.Logger.Debug(err)
+					loopError = err
+				}
+			}
 		}
 	}
 
 	return nil
 }
 
-func deleteRunner(deleteFuncName string, dFunction deleteFunc, awsSession *session.Session, filters Filter, clusterName string, logger logrus.FieldLogger, channel chan string) {
-	backoffSettings := wait.Backoff{
-		Duration: time.Second * 10,
-		Factor:   1.3,
-		Steps:    100,
+func tagMatch(filters []Filter, tags map[string]string) bool {
```

Comment:
this is OR tag match ?, seems like a good candidate for comments..

Reply:
> seems like a good candidate for comments..

We already document the `[]Filter` semantics [here][1] and [here][2].  Do you need me to repeat those for this internal helper?

[1]: https://github.com/openshift/installer/blob/v0.10.0/pkg/destroy/aws/aws.go#L29-L32
[2]: https://github.com/openshift/installer/blob/v0.10.0/pkg/destroy/aws/aws.go#L49-L64

Reply:
nit: maybe we can add a `type []Filter ORFilter` and add the function `matches` there and document the behavior. 

https://github.com/openshift/installer/blob/v0.10.0/pkg/destroy/aws/aws.go#L49-L64 is enough too.